lib/mutgen.c

```diff
@@ -187,93 +184,494 @@ mutation_model_check_validity(mutation_model_t *self)
+    const char *sep = parent_allele_length == 0 ? "" : ",";
+
+    /* Append to derived_state */
+    buff = tsk_blkalloc_get(&params->allocator, alloc_size);
+    if (buff == NULL) {
+        ret = MSP_ERR_NO_MEMORY;
+        goto out;
+    }
+    len = snprintf(buff, alloc_size, "%.*s%s%" PRId64, parent_allele_length,
+        parent_allele, sep, params->next_mutation_id);
+    if (len < 0) {
+        /* Technically this can happen. Returning TSK_ERR_IO should result
+         * in Python code checking errno. */
+        ret = TSK_ERR_IO;
+        goto out;
+    }
+    assert(len < (int) alloc_size);
+    if (params->next_mutation_id == INT64_MAX) {
+        ret = MSP_ERR_MUTATION_ID_OVERFLOW;
+        goto out;
+    }
+    params->next_mutation_id++;
+    mutation->derived_state = buff;
+    mutation->derived_state_length = (tsk_size_t) len;
+
+    /* Append to metadata */
+    buff = tsk_blkalloc_get(
+        &params->allocator, parent_metadata_length + SLIM_MUTATION_METADATA_SIZE);
+    if (buff == NULL) {
+        ret = MSP_ERR_NO_MEMORY;
+        goto out;
+    }
+    memcpy(buff, parent_metadata, parent_metadata_length);
+    copy_slim_mutation_metadata(params, buff + parent_metadata_length);
+
+    mutation->metadata = buff;
+    mutation->metadata_length
+        = (tsk_size_t)(parent_metadata_length + SLIM_MUTATION_METADATA_SIZE);
+out:
+    return ret;
+}
+
+static int
+slim_mutator_free(mutation_model_t *self)
+{
+    slim_mutator_t params = self->params.slim_mutator;
+    tsk_blkalloc_free(&params.allocator);
+    return 0;
+}
+
+/***********************
+ * Infinite alleles mutation model */
+
+static void
+infinite_alleles_print_state(mutation_model_t *self, FILE *out)
+{
+    infinite_alleles_t params = self->params.infinite_alleles;
+    fprintf(out, "infinite alleles:\n");
+    fprintf(out, "\tstart_allele:%" PRIu64 "\n", params.start_allele);
+}
+
+/* The maximum number of digits for an unsigned 64 bit integer is 20
+ * and one byte for the NULL terminator. */
+#define MAX_UINT_BUFF_SIZE 21
+
+static int
+infinite_alleles_make_allele(
+    mutation_model_t *self, char **dest, tsk_size_t *dest_length)
+{
+    int ret = 0;
+    infinite_alleles_t *params = &self->params.infinite_alleles;
+    char tmp_buff[MAX_UINT_BUFF_SIZE];
+    char *buff = NULL;
+    int err;
+    tsk_size_t len;
+
+    err = snprintf(tmp_buff, MAX_UINT_BUFF_SIZE, "%" PRIu64, params->next_allele);
+    if (err < 0) {
+        /* Technically this can happen. Returning TSK_ERR_IO should result
+         * in Python code checking errno. */
+        ret = TSK_ERR_IO;
+        goto out;
+    }
+    len = (tsk_size_t) err;
+    assert(len < MAX_UINT_BUFF_SIZE);
+
+    buff = tsk_blkalloc_get(&params->allocator, len);
+    if (buff == NULL) {
+        ret = MSP_ERR_NO_MEMORY;
+        goto out;
+    }
+    memcpy(buff, tmp_buff, len);
+    params->next_allele++;
+    *dest = buff;
+    *dest_length = (tsk_size_t) len;
```

Comment:
`len` was already cast to `tsk_size_t` on assignment - but, why not just use `len` instead of `err` throughout, as in `slim_transition_allele`?

Reply:
Good spot on the cast. We're passing ``len`` directly to blk_alloc_get as well as assigning it, so we'd have to cast it twice if we used the return value from sprintf. I thought this was the simplest and clearest. I've changed ``err`` to ``num_digits`` now though, as err was a bit misleading.